[RFC] Remove conversions which are not needed when &encoding is UTF-8

ZyX-I opened the pull request:
Problems: possible resource leak, calling convert_setup when not needed. Though &encoding is always UTF-8 now, so will drop everything in the next commit.

ZyX-I committed:
eval/decode,shada: Do not forget to clean up converters

ZyX-I committed:
shada: Initialize vimconv_T

ZyX-I committed:
eval/decode: Omit calling convert_setup for each string

Uses the same trick eval/encode does.

ZyX-I committed:
eval/*code,shada: Drop support for converting UTF-8 from/to p_enc

Not needed any longer since p_enc is always utf-8.

bfredl commented on src/nvim/eval/encode.c:
```diff
@@ -579,21 +564,7 @@ static inline int convert_to_json_string(garray_T *const gap,
-      }
-      utf_buf = tofree;
-    }
+    assert(STRCMP(p_enc, "utf-8") == 0);
```
I don't think this assert is needed here, code everywhere is free to assume that the internal encoding is UTF-8.

ZyX-I replied:
The idea was to somehow “state” that no conversions are performed because already using UTF-8, not because any string in any encoding is fine there.

bfredl replied:
isn't it better to update the comment below, which discusses the (partial) conversion to UTF-16 anyway?

ZyX-I replied:
Comments do not fail if something changed. And there are no such comments in other places where I have assert(). Also where is UTF-16 discussed?

bfredl replied:
There is nothing that can change here, that encoding is UTF-8 is a _static_ property of the codebase. `p_enc` is mutable only because that is the simplest way to allow `set enc=utf-8` and synonyms but not `set enc=whatever`, a change of that value would indicate only a failure of option validation logic, not that the encoding suddenly is not UTF-8 anymore, which is impossible. We could make this clearer maybe by changing that to use `p_enc_dummy` and making `p_enc` a constant, not by adding this assert everywhere UTF-8 is assumed.

ZyX-I replied:
I did not mean anything is going to change at runtime.

ZyX-I committed:
unittests: Do not alter p_enc in decode unit test

ZyX-I committed:
*: Fix linter errors

Drops comments `// for …` that do not pass linter for them being unmaintainable 
and fast to becoming incomplete or even incorrect.

Mention @dedmass

ZyX-I committed:
doc: Update documentation

ZyX-I committed:
*: Remove useless asserts

ZyX-I committed:
unittests: Fix linter error

ZyX-I commented:
Travis failure is single-includes.
